pkg/controller.v1/pytorch/envvar.go

```diff
@@ -89,12 +112,28 @@ func setPodEnv(obj interface{}, podTemplateSpec *corev1.PodTemplateSpec, rtype,
 			// Set elastic related environment variables.
 			podTemplateSpec.Spec.Containers[i].Env = append(
 				podTemplateSpec.Spec.Containers[i].Env, envVars...)
+		} else {
+			podTemplateSpec.Spec.Containers[i].Env = append(
+				podTemplateSpec.Spec.Containers[i].Env, corev1.EnvVar{
+					Name:  EnvNnodes,
+					Value: strconv.Itoa(int(totalReplicas)),
+				})
 		}
 	}
 
 	return nil
 }
 
+func getNprocPerNodeInt(job *kubeflowv1.PyTorchJob) int {
+	if job.Spec.NprocPerNode == nil {
+		return 1
+	}
+	if np, err := strconv.Atoi(*job.Spec.NprocPerNode); err == nil {
+		return np
+	}
+	return 1
+}
+
```

Comment:
Why do we need this? IIUC, we default `spec.NproPerNode` to `auto` if `spec.NproPerNode` is nil.

Reply:
`getNprocPerNodeInt` is a helper function to calculate world size, it will not effect the env.

When nproc_per_node set to `auto`, it means the number of process will be determinate in the user process phase, in this case, world size env will not be used.

Reply:
I see. Thanks for the clarification!
So should we add a comment to this function? And can we add validation to pkg/apis/kubeflow.org/v1/pytorch_validation.go like the following?

```go
if job.spec.NrocPerNode != nil {
   if np, err := strconv.Atoi(job.spec.NrocPerNode); err != nil && (np != "auto" || np != "CPU" || np == "GPU") {
     Error("error")
  }
}
```

Reply:
I do not think we should do that, let's say if PyTorch framework support values other than those in the list, e.g. "XPU", the operator will not work anymore.

Anyway, in my opinion, the operator should be decouple with the framework in this level.

Reply:
It makes sense.
Is there any document about supported values by PyTorch?

Reply:
Ah, I found the logic for the `nproc_per_node`: https://github.com/pytorch/pytorch/blob/26f7f470df64d90e092081e39507e4ac751f55d6/torch/distributed/run.py#L629-L658

Can you add this link to pkg/apis/kubeflow.org/v1/pytorch_types.go.

And can you add the following comment you posted to this function?

> When nproc_per_node set to auto, it means the number of process will be determinate in the user process phase, in this case, world size env will not be used.

Reply:
Others LGTM. Thanks for the big contribution.

Reply:
Done